dig/dig.go

```diff
@@ -0,0 +1,286 @@
+			}
+		}
+	}
+
+	v, err := n.value(g)
+	if err != nil {
+		return errors.Wrapf(err, "unable to resolve %v", objType)
+	}
+
+	// set the pointer value of the provided object to the instance pointer
+	objVal.Elem().Set(v)
+
+	return nil
+}
+
+type object interface {
+	// Return value of the object
+	value(g *graph) (reflect.Value, error)
+
+	// Other things that need to be present before this object can be created
+	dependencies() []interface{}
+}
+
+type graph struct {
+	fmt.Stringer
+
+	nodes map[interface{}]object
+}
+
+func newGraph() *graph {
+	return &graph{
+		nodes: make(map[interface{}]object),
+	}
+}
+
+func (g graph) String() string {
+	var b bytes.Buffer
+	for key, reg := range g.nodes {
+		b.WriteString(fmt.Sprintf("%v -> %v\n", key, reg))
+	}
+	return fmt.Sprintf("{nodes:\n%v}", b.String())
+}
+
+type node struct {
+	fmt.Stringer
+
+	obj         interface{}
+	objType     reflect.Type
+	cached      bool
+	cachedValue reflect.Value
+}
+
+// Return the earlier provided instance
+func (n node) value(g *graph) (reflect.Value, error) {
+	return reflect.ValueOf(n.obj), nil
+}
+
+func (n node) dependencies() []interface{} {
+	return nil
+}
+
+func (n node) String() string {
+	return fmt.Sprintf(
+		"(object) obj: %v, deps: nil, cached: %v",
+		n.objType,
+		n.cached,
+	)
+}
+
+type funcNode struct {
+	node
+	fmt.Stringer
+
+	constructor interface{}
+	deps        []interface{}
+}
+
+// Call the function and return the result
+func (n *funcNode) value(g *graph) (reflect.Value, error) {
+	if n.cached {
+		return n.cachedValue, nil
+	}
+
+	cv := reflect.ValueOf(n.constructor)
+	ct := reflect.TypeOf(n.constructor)
+
+	args := make([]reflect.Value, ct.NumIn(), ct.NumIn())
+	for idx := range args {
+		arg := ct.In(idx)
+		node, ok := g.nodes[arg]
+		if ok {
+			v, err := node.value(g)
+			if err != nil {
+				return reflect.Zero(n.objType), errors.Wrap(err, "dependency resolution failed")
+			}
+			args[idx] = v
+		}
+	}
+
+	v := cv.Call(args)[0]
+	n.cached = true
+	n.cachedValue = v
+
+	return v, nil
+}
+
+func (n funcNode) dependencies() []interface{} {
+	return n.deps
+}
+
+func (n funcNode) String() string {
+	return fmt.Sprintf(
+		"(function) deps: %v, type: %v, constructor: %v, cached: %v",
+		n.deps,
+		n.objType,
+		n.constructor,
+		n.cached,
+	)
+}
+
+func (g *graph) injectObject(o interface{}) error {
+	otype := reflect.TypeOf(o)
+
+	n := node{
+		obj:     o,
+		objType: otype,
+		cached:  true,
+	}
+
+	g.nodes[otype] = &n
+	return nil
+}
+
+func (g *graph) injectConstructor(c interface{}) error {
+	ctype := reflect.TypeOf(c)
+	objType := ctype.Out(0)
+
+	argc := ctype.NumIn()
+	n := funcNode{
+		deps:        make([]interface{}, argc),
+		constructor: c,
+		node: node{
+			objType: objType,
+		},
+	}
+	for i := 0; i < argc; i++ {
+		arg := ctype.In(i)
+		if arg.Kind() != reflect.Ptr && arg.Kind() != reflect.Interface {
+			return errArgKind
+		}
+
+		n.deps[i] = arg
```

Comment:
You can't check registration of arguments when constructor gets injected, because you don't control the order of when and how things get registered. We can talk about this more tomorrow, clearly this thread is already getting large